src/abstractdataframe/where.jl

```diff
@@ -0,0 +1,191 @@
+
+Return a copy of data frame `df` or parent of `gdf` containing only rows for
+which all values produced by `args` transformation in a given row is `true`. If
+`dropmissing=true` (the default) returing `missing` in `args` is allowed and
+these rows are dropped. If `dropmissing=false` an error is thrown if `missing`
+is present.
+
+Each argument passed in `args` can be either a single column selector or a
+`source_columns => function` transformation specifier following the rules
+described for [`select`](@ref).
+
+Note that as opposed to [`filter`](@ref) the `where` function works on whole columns
+and by default drops rows for which contition is false.
+
+If `view=true` a view instead of a `DataFrame` is returned.
+
+If `ungroup=true` then a `GroupedDataFrame` is returned using the same columns as
+`gdf` to group by.
+
+See also: [`where!`](@ref), [`filter`](@ref), [`filter!`](@ref)
+
+# Examples
+
+```
+julia> df = DataFrame(id=1:4, x=[true, false, true, false], y=[true, true, false, false],
+                      z=[true, true, missing, missing], v=[1, 2, 11, 12])
+4×5 DataFrame
+│ Row │ id    │ x    │ y    │ z       │ v     │
+│     │ Int64 │ Bool │ Bool │ Bool?   │ Int64 │
+├─────┼───────┼──────┼──────┼─────────┼───────┤
+│ 1   │ 1     │ 1    │ 1    │ 1       │ 1     │
+│ 2   │ 2     │ 0    │ 1    │ 1       │ 2     │
+│ 3   │ 3     │ 1    │ 0    │ missing │ 11    │
+│ 4   │ 4     │ 0    │ 0    │ missing │ 12    │
+
+julia> where(df, :x)
+2×5 DataFrame
+│ Row │ id    │ x    │ y    │ z       │ v     │
+│     │ Int64 │ Bool │ Bool │ Bool?   │ Int64 │
+├─────┼───────┼──────┼──────┼─────────┼───────┤
+│ 1   │ 1     │ 1    │ 1    │ 1       │ 1     │
+│ 2   │ 3     │ 1    │ 0    │ missing │ 11    │
+
+julia> where(df, :x, :y => ByRow(!))
+1×5 DataFrame
+│ Row │ id    │ x    │ y    │ z       │ v     │
+│     │ Int64 │ Bool │ Bool │ Bool?   │ Int64 │
+├─────┼───────┼──────┼──────┼─────────┼───────┤
+│ 1   │ 3     │ 1    │ 0    │ missing │ 11    │
+
+julia> where(df, :x, :z)
+1×5 DataFrame
+│ Row │ id    │ x    │ y    │ z     │ v     │
+│     │ Int64 │ Bool │ Bool │ Bool? │ Int64 │
+├─────┼───────┼──────┼──────┼───────┼───────┤
+│ 1   │ 1     │ 1    │ 1    │ 1     │ 1     │
+
+julia> where(df, :x, :z, dropmissing=false)
+ERROR: ArgumentError: each transformation must return a vector whose eltype is subtype of Bool
+
+julia> where(groupby(df, :y), :v => x -> x .> minimum(x), dropmissing=false)
+2×5 DataFrame
+│ Row │ id    │ x    │ y    │ z       │ v     │
+│     │ Int64 │ Bool │ Bool │ Bool?   │ Int64 │
+├─────┼───────┼──────┼──────┼─────────┼───────┤
+│ 1   │ 2     │ 0    │ 1    │ 1       │ 2     │
+│ 2   │ 4     │ 0    │ 0    │ missing │ 12    │
+```
+"""
+function where(df::AbstractDataFrame, @nospecialize(args...);
+               dropmissing::Bool=true, view::Bool=false)
+    row_selector = _get_where_conditions(df, args, dropmissing)
+    return view ? Base.view(df, row_selector, :) : df[row_selector, :]
+end
```

Comment:
Probably worth using `@inline` to avoid the type instability, with `@noinline` in `_get_where_conditions` (with inference tests to prevent regressions). Same for the `GroupedDataFrame` method.

Reply:
OK - I have added `@inline`.

Reply:
The code is type unstable even with `@inline` and `@noinline` in `_get_where_conditions`. Do you think it can be resolved somehow? (and do you think it is so problematic given `DataFrame` is type unstable anyway?)

Reply:
That's weird. We use this kind of pattern elsewhere and it works, right? Is `row_selector` correctly inferred?

Reply:
@nalimilan - given we add `ungroup` kwarg I am not sure we can get type stability here, do we? (as now we have yet another option of the return value related to the second kwarg)

Reply:
`ungroup` shouldn't really make things worse as long as the function is inlined and the value of arguments know at compile time. But if that doesn't work in the first place...

Reply:
This is what I have now:
```
julia> @code_warntype subset(DataFrame(a=1:10), :a => ByRow(<(4)), view=false, skipmissing=false)
Variables
  #unused#::Core.Compiler.Const(DataFrames.var"#subset##kw"(), false)
  @_2::NamedTuple{(:view, :skipmissing),Tuple{Bool,Bool}}
  @_3::Core.Compiler.Const(DataFrames.subset, false)
  df::DataFrame
  args::Tuple{Pair{Symbol,ByRow{Base.Fix2{typeof(<),Int64}}}}
  skipmissing::Bool
  view::Bool
  @_8::Bool
  @_9::Bool

Body::Any
1 ── %1  = Base.haskey(@_2, :skipmissing)::Core.Compiler.Const(true, false)
│          %1
│    %3  = Base.getindex(@_2, :skipmissing)::Bool
│    %4  = (%3 isa DataFrames.Bool)::Core.Compiler.Const(true, false)
│          %4
└───       goto #3
2 ──       Core.Compiler.Const(:(%new(Core.TypeError, Symbol("keyword argument"), :skipmissing, DataFrames.Bool, %3)), false)
└───       Core.Compiler.Const(:(Core.throw(%7)), false)
3 ┄─       (@_8 = %3)
└───       goto #5
4 ──       Core.Compiler.Const(:(@_8 = false), false)
5 ┄─       (skipmissing = @_8)
│    %13 = Base.haskey(@_2, :view)::Core.Compiler.Const(true, false)
│          %13
│    %15 = Base.getindex(@_2, :view)::Bool
│    %16 = (%15 isa DataFrames.Bool)::Core.Compiler.Const(true, false)
│          %16
└───       goto #7
6 ──       Core.Compiler.Const(:(%new(Core.TypeError, Symbol("keyword argument"), :view, DataFrames.Bool, %15)), false)
└───       Core.Compiler.Const(:(Core.throw(%19)), false)
7 ┄─       (@_9 = %15)
└───       goto #9
8 ──       Core.Compiler.Const(:(@_9 = false), false)
9 ┄─       (view = @_9)
│    %25 = (:skipmissing, :view)::Core.Compiler.Const((:skipmissing, :view), false)
│    %26 = Core.apply_type(Core.NamedTuple, %25)::Core.Compiler.Const(NamedTuple{(:skipmissing, :view),T} where T<:Tuple, false)
│    %27 = Base.structdiff(@_2, %26)::Core.Compiler.Const(NamedTuple(), false)
│    %28 = Base.pairs(%27)::Core.Compiler.Const(Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}(), false)
│    %29 = Base.isempty(%28)::Core.Compiler.Const(true, false)
│          %29
└───       goto #11
10 ─       Core.Compiler.Const(:(Core.tuple(@_2, @_3, df)), false)
└───       Core.Compiler.Const(:(Core._apply_iterate(Base.iterate, Base.kwerr, %32, args)), false)
11 ┄ %34 = Core.tuple(skipmissing, view, @_3, df)::Core.Compiler.PartialStruct(Tuple{Bool,Bool,typeof(subset),DataFrame}, Any[Bool, Bool, Core.Compiler.Const(DataFrames.subset, false), DataFrame])
│    %35 = Core._apply_iterate(Base.iterate, DataFrames.:(var"#subset#406"), %34, args)::Any
└───       return %35

julia> f() = subset(DataFrame(a=1:10), :a => ByRow(<(4)), view=false, skipmissing=false)
f (generic function with 1 method)

julia> @code_warntype f()
Variables
  #self#::Core.Compiler.Const(f, false)

Body::Any
1 ─ %1  = (1:10)::Core.Compiler.Const(1:10, false)
│   %2  = (:a,)::Core.Compiler.Const((:a,), false)
│   %3  = Core.apply_type(Core.NamedTuple, %2)::Core.Compiler.Const(NamedTuple{(:a,),T} where T<:Tuple, false)
│   %4  = Core.tuple(%1)::Core.Compiler.Const((1:10,), false)
│   %5  = (%3)(%4)::Core.Compiler.Const((a = 1:10,), false)
│   %6  = Core.kwfunc(Main.DataFrame)::Core.Compiler.Const(Core.var"#Type##kw"(), false)
│   %7  = (%6)(%5, Main.DataFrame)::DataFrame
│   %8  = (<)(4)::Core.Compiler.Const(Base.Fix2{typeof(<),Int64}(<, 4), false)
│   %9  = Main.ByRow(%8)::Core.Compiler.Const(ByRow{Base.Fix2{typeof(<),Int64}}(Base.Fix2{typeof(<),Int64}(<, 4)), false)
│   %10 = (:a => %9)::Core.Compiler.Const(:a => ByRow{Base.Fix2{typeof(<),Int64}}(Base.Fix2{typeof(<),Int64}(<, 4)), false)
│   %11 = (:view, :skipmissing)::Core.Compiler.Const((:view, :skipmissing), false)
│   %12 = Core.apply_type(Core.NamedTuple, %11)::Core.Compiler.Const(NamedTuple{(:view, :skipmissing),T} where T<:Tuple, false)
│   %13 = Core.tuple(false, false)::Core.Compiler.Const((false, false), false)
│   %14 = (%12)(%13)::Core.Compiler.Const((view = false, skipmissing = false), false)
│   %15 = Core.kwfunc(Main.subset)::Core.Compiler.Const(DataFrames.var"#subset##kw"(), false)
│   %16 = (%15)(%14, Main.subset, %7, %10)::Any
└──       return %16
```
(not passing kwargs does not change this)

Maybe the reason is that we have a vararg function with `@nospecialize`?

Reply:
removing `@nospecialize` did not change anything.